8278114: New addnode ideal optimization: converting "x + x" into "x << 1"

CptGit opened the pull request:
A new ideal optimization can be introduced for addnode: converting "x + x" into "x << 1".

```
// Convert "x + x" into "x << 1"
if (in1 == in2) {
  return new LShiftINode(in1, phase->intcon(1));
}
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278114](https://bugs.openjdk.java.net/browse/JDK-8278114): New addnode ideal optimization: converting "x + x" into "x << 1"


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 306c95a3caa9815daecde40743d138283f7c6c67


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6675/head:pull/6675` \
`$ git checkout pull/6675`

Update a local copy of the PR: \
`$ git checkout pull/6675` \
`$ git pull https://git.openjdk.java.net/jdk pull/6675/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6675`

View PR using the GUI difftool: \
`$ git pr show -t 6675`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6675.diff">https://git.openjdk.java.net/jdk/pull/6675.diff</a>

</details>

CptGit commented:
@merykitty If you get a chance could you please take a quick look at this PR? I was wondering if this conversion makes sense. I am concerned if this obvious optimization has been done somewhere else which I was not able to find. Thank you very much.

merykitty commented:
Tbh I don't find this transformation necessary, `x + x` is a very cheap operation and is generally easier for the optimiser to work with than `x << 1`.
Cheers.

cl4es commented:
> Tbh I don't find this transformation necessary, `x + x` is a very cheap operation and is generally easier for the optimiser to work with than `x << 1`. Cheers.

I agree this looks to be of dubious value on the face of it. A microbenchmark to prove it's beneficial in some scenario feels like a requirement here.

A targeted microbenchmark could explore if we already do or could better allow constant folding of expressions that does subsequent shifts, e.g. by turning `(x + x) << 3` into `x << 4`.

CptGit commented:
> > Tbh I don't find this transformation necessary, `x + x` is a very cheap operation and is generally easier for the optimiser to work with than `x << 1`. Cheers.
> 
> I agree this looks to be of dubious value on the face of it. A microbenchmark to prove it's beneficial in some scenario feels like a requirement here.
> 
> A targeted microbenchmark could explore if we already do or could better allow constant folding of expressions that does subsequent shifts, e.g. by turning `(x + x) << 3` into `x << 4`.

Thank you both for commenting and suggestions! @merykitty @cl4es 

I moved the transformation into LShiftNode, i.e., to convert `(x + x) << 3` into `x << 4`, and I observed a 5% performance improvement (percentage of average time reduced) via the microbenchmark I write. I am not sure how much improvement we usually expect for a transformation, is this one currently a beneficial transformation? Thank you!

cl4es commented on test/micro/org/openjdk/bench/vm/compiler/AddIdeal_XPlusX_LShiftC.java:
```diff
@@ -0,0 +1,103 @@
+@State(Scope.Thread)
+@Warmup(iterations = 20, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 20, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 3 , jvmArgsAppend = {"-XX:-TieredCompilation"})
```
Is `-TieredCompilation` necessary to demonstrate an effect? Settings that are not strictly necessary - such as tuning - should generally be avoided (someone might want to set up separate runs with `TieredCompilation` enabled and disabled on a higher level, for example)

cl4es commented:
> Thank you both for commenting and suggestions! @merykitty @cl4es
> 
> I moved the transformation into LShiftNode, i.e., to convert `(x + x) << 3` into `x << 4`, and I observed a 5% performance improvement (percentage of average time reduced) via the microbenchmark I write. I am not sure how much improvement we usually expect for a transformation, is this one currently a beneficial transformation? Thank you!

For such a small win you might be looking at noise and should inspect the generated assembly to see that there's any real difference in the emitted code. 

Running your `test` microbenchmark using `-prof perfasm` on my Linux workstation the relevant code generated for the `helper` method with your patch looks like this:

```
  4.40%     0x00007f2b0d6d720c:   mov    %esi,%eax
            0x00007f2b0d6d720e:   and    $0x1fffff,%eax               ;*iushr {reexecute=0 rethrow=0 return_oop=0}
                                                                      ; - org.openjdk.bench.vm.compiler.AddIdeal_XPlusX_LShiftC::helper@8 (line 88)
```

i.e., the `((x + x) << 10) >> 11` is fully transformed down into `x & 0x1fffff`, which seem like it could be a pretty optimal result for this code (zeroes out the top 11 bits).

Baseline build generates this:
```
  4.46%      0x00007f93e16d988c:   add    %esi,%esi
             0x00007f93e16d988e:   shl    $0xa,%esi
             0x00007f93e16d9891:   mov    %esi,%eax
             0x00007f93e16d9893:   shr    $0xb,%eax                    ;*iushr {reexecute=0 rethrow=0 return_oop=0}
                                                                       ; - org.openjdk.bench.vm.compiler.AddIdeal_XPlusX_LShiftC::helper@8 (line 88)
```

Which is a pretty direct transformation of the Java code into assembly. So it seems your patch _does_ enable more optimal transformations!

Only a few percent of the benchmark time seem to be spent in the relevant code, though. As written I can't establish a significant result between a baseline VM and a patched one as the variance from run to run is more than a few percent. A narrowed down benchmark that only tests a single value seem to better demonstrate the low-level effect:

```
    @Benchmark
    public int testInt() {
        return helper(ints_a[4711]);
    }
    
    @Benchmark
    public long testLong() {
        return helper(longs_a[4711]);
    }
```

Not using the hand-rolled `sink` blackhole means we rely on JMH infra, which can since recently cooperate with the VM to further remove overhead if you run with `-Djmh.blackhole.autoDetect=true` (soon enabled by default). I also removed `-TieredCompilation` from the `@Fork` and get slightly better and stable scores overall, though a bit longer warmup is needed.

Baseline:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt   15  8.925 ± 0.019  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt   15  8.969 ± 0.107  ns/op
```
Patch:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt    4  7.750 ± 1.007  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt    4  7.681 ± 0.234  ns/op
```

A ~1.2ns improvement - or 1.15x speed-up after removing as much infrastructural overhead as possible. I think this looks in line with the observable change in the generated assembly (while we get rid of more than one instruction, they are likely pretty nicely pipelined on my Intel CPU). Still I think this looks like a reasonable improvement for a relatively straightforward change to C2.

CptGit commented:
Thank you @cl4es so much for taking time to review my pull request and deeply investigate the microbenchmark! I learnt a lot on writing better JMH microbenchmarks by reading your comment.
> > Thank you both for commenting and suggestions! @merykitty @cl4es
> > I moved the transformation into LShiftNode, i.e., to convert `(x + x) << 3` into `x << 4`, and I observed a 5% performance improvement (percentage of average time reduced) via the microbenchmark I write. I am not sure how much improvement we usually expect for a transformation, is this one currently a beneficial transformation? Thank you!
> 
> For such a small win you might be looking at noise and should inspect the generated assembly to see that there's any real difference in the emitted code.
> 
Great to know we can print assembly! Thanks.
> Running your `test` microbenchmark using `-prof perfasm` on my Linux workstation the relevant code generated for the `helper` method with your patch looks like this:
> 
> ```
>   4.40%     0x00007f2b0d6d720c:   mov    %esi,%eax
>             0x00007f2b0d6d720e:   and    $0x1fffff,%eax               ;*iushr {reexecute=0 rethrow=0 return_oop=0}
>                                                                       ; - org.openjdk.bench.vm.compiler.AddIdeal_XPlusX_LShiftC::helper@8 (line 88)
> ```
> 
> i.e., the `((x + x) << 10) >> 11` is fully transformed down into `x & 0x1fffff`, which seem like it could be a pretty optimal result for this code (zeroes out the top 11 bits).
> 
> Baseline build generates this:
> 
> ```
>   4.46%      0x00007f93e16d988c:   add    %esi,%esi
>              0x00007f93e16d988e:   shl    $0xa,%esi
>              0x00007f93e16d9891:   mov    %esi,%eax
>              0x00007f93e16d9893:   shr    $0xb,%eax                    ;*iushr {reexecute=0 rethrow=0 return_oop=0}
>                                                                        ; - org.openjdk.bench.vm.compiler.AddIdeal_XPlusX_LShiftC::helper@8 (line 88)
> ```
> 
> Which is a pretty direct transformation of the Java code into assembly. So it seems your patch _does_ enable more optimal transformations!
> 
> Only a few percent of the benchmark time seem to be spent in the relevant code, though. As written I can't establish a significant result between a baseline VM and a patched one as the variance from run to run is more than a few percent. A narrowed down benchmark that only tests a single value seem to better demonstrate the low-level effect:
> 
> ```
>     @Benchmark
>     public int testInt() {
>         return helper(ints_a[4711]);
>     }
>     
>     @Benchmark
>     public long testLong() {
>         return helper(longs_a[4711]);
>     }
> ```
> 
Thanks for point out this new feature. I did use `Blackhole` but found it brought too much overhead from infra so I referred to JMH examples for a sink method. I'll start using `Backhole` with `-Djmh.blackhole.autoDetect=true` from now on.
> Not using the hand-rolled `sink` blackhole means we rely on JMH blackholes. Those traditionally had a bit more overhead than what you get with `sink`, but can since recently cooperate with the VM to remove overhead almost completely if you run with `-Djmh.blackhole.autoDetect=true` (soon enabled by default). I also removed `-TieredCompilation` from the `@Fork` and get slightly better and stable scores overall, though a bit longer warmup is needed.
> 
> Baseline:
> 
> ```
> Benchmark                         Mode  Cnt  Score   Error  Units
> AddIdeal_XPlusX_LShiftC.testInt   avgt   15  8.925 ± 0.019  ns/op
> AddIdeal_XPlusX_LShiftC.testLong  avgt   15  8.969 ± 0.107  ns/op
> ```
> 
> Patch:
> 
> ```
> Benchmark                         Mode  Cnt  Score   Error  Units
> AddIdeal_XPlusX_LShiftC.testInt   avgt   15  7.650 ± 0.024  ns/op
> AddIdeal_XPlusX_LShiftC.testLong  avgt   15  7.655 ± 0.087  ns/op
> ```
> 
If you get a chance, could you please share the modifications you made to the microbenchmark, if you made other changes than removing `-TieredCompilation` and testing only a single value, so I can test on my end, for more data points? Also, the examples in JMH repository seem out-of-date (e.g., the `Blackhole` autodetect feature) I think your example will be a good study material for me. Thank you.
> A ~1.2ns improvement - or 1.15x speed-up after removing as much infrastructural overhead as possible. I think this looks in line with the observable change in the generated assembly (while we get rid of more than one instruction, they are likely pretty nicely pipelined on my Intel CPU). I think this looks like a reasonable improvement for such a straightforward change to C2.

cl4es commented:
> Thank you @cl4es so much for taking time to review my pull request and deeply investigate the microbenchmark! I learnt a lot on writing better JMH microbenchmarks by reading your comment.

My pleasure. I was intrigued since you picked up on my suggestion to enable folding of constant shift transforms so fast and wanted to see for myself that it ended up compiling down to something more compact/optimal.

> If you get a chance, could you please share the modifications you made to the microbenchmark, if you made other changes than removing `-TieredCompilation` and testing only a single value, so I can test on my end, for more data points? Also, the examples in JMH repository seem out-of-date (e.g., the `Blackhole` autodetect feature) I think your example will be a good study material for me. Thank you.

No other code changes, just added in those two simplified benchmarks and commented out the existing ones. I did tweak the command line to print using `ns` (you can change the annotation to get the same result) and picked the smallest number of iterations that still gave stable results on my machine while trying things out: `-i 15 -wi 15 -tu ns`.

CptGit commented:
> No other code changes, just added in those two simplified benchmarks and commented out the existing ones. I did tweak the command line to print using `ns` (you can change the annotation to get the same result) and picked the smallest number of iterations that still gave stable results on my machine while trying things out: `-i 15 -wi 15 -tu ns`.

Is there special configuration you were using? I was not able to reproduce the similar results.

This is when I used 20 iterations.
Baseline:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt   60  3.846 ? 0.164  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt   60  3.332 ? 0.008  ns/op
```

Patch:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt   60  3.743 ? 0.245  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt   60  3.664 ? 0.224  ns/op
```

Then I tried 100 interations to reduce noise.
Baseline:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt  300  3.566 ? 0.068  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt  300  3.310 ? 0.004  ns/op
```

Patch:
```
Benchmark                         Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.testInt   avgt  300  3.724 ? 0.103  ns/op
AddIdeal_XPlusX_LShiftC.testLong  avgt  300  3.605 ? 0.086  ns/op
```

I were not able to print assembly on my end because of some errors from `perf` which I do not have permission to configure, but I checked native coverage and it showed this transformation did happen when the microbenchmark was running.

merykitty commented:
Alternatively, you could use `-XX:CompileCommand=print,*AddIdeal_XPlusX_LShiftC.testInt` which will print every detail about compilation including opto assembly. In this case, you could use `-XX:-TieredCompilation` to not receive details regarding C1 compiled method. Note that without `hsdis` the JVM will only output machine code which is very hard to analyse.

merykitty commented:
The reason you don't see any shift instruction is that `x << 11 >>> 11` is transformed into `x & 0x1FFFFF` which is the instruction you see in `0x00007f97b9418dae:   and    $0x1fffff,%eax`. As you can see for these kinds of microbenchmarks the cost of calling functions dominates the cost of the actual transformation, you can mitigate this by pulling the helper method out to be the benchmark (I don't see the reason to have a separate helper method, either), and putting the operation in a loop that has the results being sunk in a compiler blackhole (full blackhole or a non-inlined sink method won't work as effective in this case simply because it is also a function call) in each iteration.

CptGit commented:
Thank you very much! I adapt the microbenmark according to your advice and I am able to observed a notable difference now! 

I ran `make test TEST="micro:LShiftIdeal_XPlusX_LShiftC" MICRO="JAVA_OPTIONS=-Djmh.blackhole.mode=COMPILER" CONF_NAME="cov"` and this is the result:

Baseline:
```
Benchmark                             Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.baselineInt   avgt   60  1.457 ± 0.006  ns/op
AddIdeal_XPlusX_LShiftC.baselineLong  avgt   60  1.453 ± 0.004  ns/op
AddIdeal_XPlusX_LShiftC.testInt       avgt   60  2.521 ± 0.010  ns/op
AddIdeal_XPlusX_LShiftC.testLong      avgt   60  2.518 ± 0.009  ns/op
```

Patch:
```
Benchmark                             Mode  Cnt  Score   Error  Units
AddIdeal_XPlusX_LShiftC.baselineInt   avgt   60  1.455 ± 0.005  ns/op
AddIdeal_XPlusX_LShiftC.baselineLong  avgt   60  1.453 ± 0.005  ns/op
AddIdeal_XPlusX_LShiftC.testInt       avgt   60  1.455 ± 0.006  ns/op
AddIdeal_XPlusX_LShiftC.testLong      avgt   60  1.677 ± 0.005  ns/op
```

> As you can see for these kinds of microbenchmarks the cost of calling functions dominates the cost of the actual transformation, you can mitigate this by pulling the helper method out to be the benchmark (I don't see the reason to have a separate helper method, either), and putting the operation in a loop that has the results being sunk in a compiler blackhole (full blackhole or a non-inlined sink method won't work as effective in this case simply because it is also a function call) in each iteration.

kelthuzadx commented:
More thought: To see what GCC-O2 does for the same code.

merykitty commented:
All gcc, clang and msvc seem to perform this transformation, [godbolt](https://godbolt.org/z/sYYxrTjPY).
Cheers.

CptGit commented:
Hope everybody had a good break and happy new year!

@cl4es if you get a chance can you make a review again? Thank you.

vnkozlov reviewed (approved):
Looks good to me.

vnkozlov commented:
You still need approval from @cl4es

CptGit commented:
/reviewers 2

cl4es reviewed (approved):
Looks good. The further microbenchmark improvements suggested by @merykitty makes a lot of sense given the context. 

The comment about compiler blackhole is becoming a bit redundant with the upgrade to JMH 1.34 (#6955), which enables the compiler assisted blackholes by default.

vnkozlov reviewed (changes requested):
I forgot to request to update copyright year to 2022 in changed and new files headers.

CptGit commented:
> I forgot to request to update copyright year to 2022 in changed files headers.

Good catch! Also I addressed the redundant comment about compiler blackhole in the microbenchmark.

vnkozlov reviewed (approved):
Good. And my testing passed.

CptGit commented:
> Good. And my testing passed.
Cool, thanks for checking. I think I can integrate now?

vnkozlov commented:
> Cool, thanks for checking. I think I can integrate now?

Yes. And I will sponsor it.

CptGit commented:
/integrate
